File: apps/website/src/lib/editor-ffi/required-surface-tiles.ts
import type { Rect } from '@typie/editor-ffi/browser';

const TILE_SIZE = 512;
const MAX_TILES = 128;

// Logical page regions become a stable device-pixel grid shared with the Rust rasterizer.
export function requiredSurfaceTiles(width: number, height: number, scale: number, regions: readonly Rect[]): number[] {
  const pixelWidth = Math.round(width * scale);
  const pixelHeight = Math.round(height * scale);
  if (!Number.isFinite(pixelWidth) || !Number.isFinite(pixelHeight) || !Number.isFinite(scale) || scale <= 0) return [];
  const tiles = new Map<string, number[]>();
  for (const region of regions) {
    if (![region.x, region.y, region.width, region.height].every(Number.isFinite)) continue;
    const left = Math.max(0, Math.min(pixelWidth, Math.floor(region.x * scale)));
    const top = Math.max(0, Math.min(pixelHeight, Math.floor(region.y * scale)));
    const right = Math.max(0, Math.min(pixelWidth, Math.ceil((region.x + region.width) * scale)));
    const bottom = Math.max(0, Math.min(pixelHeight, Math.ceil((region.y + region.height) * scale)));
    if (right <= left || bottom <= top) continue;
    const firstColumn = Math.floor(left / TILE_SIZE);
    const lastColumn = Math.floor((right - 1) / TILE_SIZE);
    const firstRow = Math.floor(top / TILE_SIZE);
    const lastRow = Math.floor((bottom - 1) / TILE_SIZE);
    if ((lastColumn - firstColumn + 1) * (lastRow - firstRow + 1) > MAX_TILES) throw new Error('Visible surface exceeds tile budget');
    for (let row = firstRow; row <= lastRow; row++) {
      for (let column = firstColumn; column <= lastColumn; column++) {
        const x = column * TILE_SIZE;
        const y = row * TILE_SIZE;
        tiles.set(`${column},${row}`, [x, y, Math.min(x + TILE_SIZE, pixelWidth), Math.min(y + TILE_SIZE, pixelHeight)]);
        if (tiles.size > MAX_TILES) throw new Error('Visible surface exceeds tile budget');
      }
    }
  }
  return [...tiles.values()].toSorted((a, b) => a[1] - b[1] || a[0] - b[0]).flat();
}

File: apps/mobile/compose/src/commonMain/kotlin/co/typie/editor/RequiredSurfaceTiles.kt
package co.typie.editor

import androidx.compose.ui.geometry.Rect
import androidx.compose.ui.unit.IntRect
import kotlin.math.ceil
import kotlin.math.floor
import kotlin.math.roundToInt

internal const val EditorTileSize = 512
internal const val EditorTileGutter = 1
internal const val MaximumSurfaceTiles = 128

/** Regions are page-local logical coordinates; the grid is fixed in raster pixels. */
internal fun requiredSurfaceTiles(
  width: Double,
  height: Double,
  scaleFactor: Double,
  regions: List<Rect>?,
): List<IntRect> {
  if (!width.isFinite() || !height.isFinite() || !scaleFactor.isFinite() || scaleFactor <= 0.0)
    return emptyList()
  val pixelWidth = (width * scaleFactor).roundToInt().coerceAtLeast(1)
  val pixelHeight = (height * scaleFactor).roundToInt().coerceAtLeast(1)
  val requested = regions ?: listOf(Rect(0f, 0f, width.toFloat(), height.toFloat()))
  val tiles = mutableSetOf<IntRect>()
  for (region in requested) {
    if (
      !region.left.isFinite() ||
        !region.top.isFinite() ||
        !region.right.isFinite() ||
        !region.bottom.isFinite()
    )
      continue
    val left =
      floor(region.left.toDouble() * scaleFactor).coerceIn(0.0, pixelWidth.toDouble()).toInt()
    val top =
      floor(region.top.toDouble() * scaleFactor).coerceIn(0.0, pixelHeight.toDouble()).toInt()
    val right =
      ceil(region.right.toDouble() * scaleFactor).coerceIn(0.0, pixelWidth.toDouble()).toInt()
    val bottom =
      ceil(region.bottom.toDouble() * scaleFactor).coerceIn(0.0, pixelHeight.toDouble()).toInt()
    if (right <= left || bottom <= top) continue
    val columns = left / EditorTileSize..(right - 1) / EditorTileSize
    val rows = top / EditorTileSize..(bottom - 1) / EditorTileSize
    require(columns.count().toLong() * rows.count() <= MaximumSurfaceTiles) {
      "Visible surface exceeds tile budget"
    }
    for (row in rows) for (column in columns) {
      val x = column * EditorTileSize
      val y = row * EditorTileSize
      tiles +=
        IntRect(
          x,
          y,
          minOf(x.toLong() + EditorTileSize, pixelWidth.toLong()).toInt(),
          minOf(y.toLong() + EditorTileSize, pixelHeight.toLong()).toInt(),
        )
      require(tiles.size <= MaximumSurfaceTiles) { "Visible surface exceeds tile budget" }
    }
  }
  return tiles.sortedWith(compareBy<IntRect> { it.top }.thenBy { it.left })
}
